### PR TITLE
Add support for measuring, combining, and getting bounds of Path objects 

### PR DESCRIPTION
**Pull request by dnfield**
I've closed #4731 in favor of this (that PR got the CLA bot very confused and Travis was starting to get off kilter with it as well).

Changes from the end of this PR to this PR:

1. Fixed issue with iteration where first valid path would have been skipped by a `for..in` loop
1. Remove `new Path.from(` in `computeMetrics` as it was determined the Skia team handles this internally
1. Updated documentation comments

The immediate use case for these changes is to support a Flutter/Dart native port of Lottie (from Android).  Initial issue raised at flutter/flutter#10428

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -1558,21 +1586,40 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// If matrix4 is specified, the path will be transformed by this matrix
+  /// (after the matrix is translated by the given offset).
+  void extendWithPath(Path path, Offset offset, {Float64List matrix4}) {
     assert(path != null); // path is checked on the engine side
     assert(_offsetIsValid(offset));
+
+    if (matrix4 != null) {
+      assert(_matrix4IsValid(matrix4));
+      _extendWithPathAndMatrix(path, offset.dx, offset.dy, matrix4);
+    }
```
Comment (Hixie) [inline lib/ui/painting.dart]: you need an else or a return here

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -1558,21 +1586,40 @@ class Path extends NativeFieldWrapperClass2 {
...
     _extendWithPath(path, offset.dx, offset.dy);
```
Comment (Hixie) [inline lib/ui/painting.dart]: given that you've added a way to enumerate the path, this would be a perfect candidate for something to test. :-)

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -1607,12 +1654,160 @@ class Path extends NativeFieldWrapperClass2 {
...
+
+  PathMetricIterator._(this._pathMetric);
+
+  @override
+  PathMetric get current => _pathMetric;
+
+  @override
+  bool moveNext() {
+    // PathMetric isn't a normal iterable - it's already initialized to its
+    // first Path.  Should only call _moveNext when done with the first one.
```
Comment (Hixie) [inline lib/ui/painting.dart]: this violates the Iterator contract. We should make sure we obey the contact precisely.

Comment (Hixie) [reply lib/ui/painting.dart]: (which is easy, just have current return `_firstTime ? null : _pathMetric`)

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -1607,12 +1654,160 @@ class Path extends NativeFieldWrapperClass2 {
...
+
+  /// Given a start and stop distance, return the intervening segment(s).
+  /// 
+  /// startD and stopD are pinned to legal values (0..[length])
+  /// Returns null if the segment is 0 length or startD > stopD.
+  /// Begin the segment with a moveTo if startWithMoveTo is true.
+  Path extractPath(double startD, double endD, bool startWithMoveTo) native 'PathMeasure_getSegment';
+
+  /// Returns true if the path is closed.
+  bool get closed native 'PathMeasure_isClosed';
```
Comment (Hixie) [inline lib/ui/painting.dart]: let's call this isClosed (sorry if i'm contradicting an earlier suggestion)

File: tools/licenses/pubspec.lock [hunk 5]
```diff
@@ -1,61 +1,53 @@
...
   typed_data:
-    dependency: transitive
     description:
       name: typed_data
       url: "https://pub.dartlang.org"
     source: hosted
     version: "1.1.5"
 sdks:
-  dart: ">=1.21.0 <=2.0.0-edge.0d5cf900b021bf5c9fa593ffa12b15bcd1cc5fe0"
+  dart: ">=1.21.0 <2.0.0"
```
Comment (Hixie) [inline tools/licenses/pubspec.lock]: revert the changes to this file before you submit

File: lib/ui/painting.dart (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting.dart]: The enumeration is only possible if there are multiple disconnected segments.  That could certainly be tested.  Am I right in believing those tests would go in flutter/flutter?

**Comment by Hixie** [comment]
The code in this patch is very close. I just want to get clarify from the skia folks about how much we care about the combine method failing first. It sounds like it might be an extremely rare thing that we don't need to worry about.

Once that's figured out I'll do a review pass for documentation and style.

File: lib/ui/painting.dart (hunk 2 above)
Comment (Hixie) [reply lib/ui/painting.dart]: What I meant was that the extendWithPath method would be a good candidate to test, e.g. here you would have tested calling the method with a path that has two segments, and given a matrix, and verified that the result had the right number of segments -- without the return or else, it wouldn't).

But yes, everything in general should be tested. For this kind of thing, the appropriate place is:
https://github.com/flutter/engine/tree/master/testing/dart

**Comment by Hixie** [comment]
Ok I spoke more to the Skia folks and apparently returning false from Op is considered a rare edge case and usually indicates invalid inputs (like NaN). So I think rather than `orElse` we should just make it throw a StateError if the operation returns false.

File: lib/ui/painting.dart [hunk 6]
```diff
@@ -39,6 +39,13 @@ bool _offsetIsValid(Offset offset) {
   return true;
 }
 
+bool _matrix4IsValid(Float64List matrix4) {
+  assert(matrix4 != null, 'Matrix4 argument was null.');
+  if (matrix4.length != 16)
+    throw new ArgumentError('"matrix4" must have 16 entries.');
```
Comment (Hixie) [inline lib/ui/painting.dart]: Since you only call this from asserts, maybe just make this an assert (like the line above) rather than throwing an explicit exception

File: lib/ui/painting.dart [hunk 7]
```diff
@@ -1340,6 +1347,18 @@ enum PathFillType {
   evenOdd,
 }
 
+/// Path operation enums
```
Comment (Hixie) [inline lib/ui/painting.dart]: here and below, all sentences in dartdocs should end with a period.

Comment (Hixie) [inline lib/ui/painting.dart]: "Path operation enums" doesn't really describe what this does.

Maybe:

```
/// Strategies for combining paths.
///
/// See also:
///
///  * [Path.combine], which uses this enum to decide how to combine two paths.
```

File: lib/ui/painting.dart [hunk 8]
```diff
@@ -1340,6 +1347,18 @@ enum PathFillType {
...
+/// 
+/// For use with [Path.combine]
+/// Must be kept in sync with SkPathOp
```
Comment (Hixie) [inline lib/ui/painting.dart]: This line should be a comment (`//`), not a dartdoc (`///`), so that it doesn't appear in the docs.

File: lib/ui/painting.dart [hunk 9]
```diff
@@ -1340,6 +1347,18 @@ enum PathFillType {
...
+enum PathOperation {
+    difference,         //!< subtract the op path from the first path
```
Comment (Hixie) [inline lib/ui/painting.dart]: These should have documentation on the line before them, with a blank line between them, and indented by two not four. Also, the descriptions should be complete sentences, maybe with examples.

```
  /// Subtract the second path from the first path.
  ///
  /// For example, if the first path is a circle, and the second path is a smaller
  /// circle in the center of the first, then the result is a doughnut. 
  difference,

  /// Intersect the two paths.
  ///
  /// For example, if the paths are a long horizontal rectangle intersecting a long
  /// vertical rectangle with the same width, the result will be a square, where they
  /// would overlap.
  intersect,
```
...and so on.

File: lib/ui/painting.dart [hunk 10]
```diff
@@ -1362,6 +1381,15 @@ class Path extends NativeFieldWrapperClass2 {
   Path() { _constructor(); }
   void _constructor() native 'Path_constructor';
 
+  /// Creates a new [Path] object from src path
```
Comment (Hixie) [inline lib/ui/painting.dart]: Create a copy of another path.

File: lib/ui/painting.dart [hunk 11]
```diff
@@ -1362,6 +1381,15 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// 
+  /// This copy is fast and does not require additional memory unless the src or
+  /// new path are mutated.
+  factory Path.from(Path src) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: `src` -> `source` (we avoid abbreviations)

File: lib/ui/painting.dart [hunk 12]
```diff
@@ -1362,6 +1381,15 @@ class Path extends NativeFieldWrapperClass2 {
   Path() { _constructor(); }
   void _constructor() native 'Path_constructor';
 
+  /// Creates a new [Path] object from src path
+  /// 
+  /// This copy is fast and does not require additional memory unless the src or
```
Comment (Hixie) [inline lib/ui/painting.dart]: unless either the source path or the path returned by this constructor are modified.

File: lib/ui/painting.dart [hunk 13]
```diff
@@ -1558,21 +1586,41 @@ class Path extends NativeFieldWrapperClass2 {
 
   /// Adds a new subpath that consists of the given path offset by the given
   /// offset.
-  void addPath(Path path, Offset offset) {
+  /// 
+  /// If matrix4 is specified, the path will be transformed by this matrix
```
Comment (Hixie) [inline lib/ui/painting.dart]: Put argument names (e.g. matrix4 here) in backticks.

File: lib/ui/painting.dart [hunk 14]
```diff
@@ -1558,21 +1586,41 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// (after the matrix is translated by the given offset).
```
Comment (Hixie) [inline lib/ui/painting.dart]: remove the parentheses, the information there is very important and should be part of the main flow of text.

File: lib/ui/painting.dart [hunk 15]
```diff
@@ -1558,21 +1586,41 @@ class Path extends NativeFieldWrapperClass2 {
...
+  void addPath(Path path, Offset offset, {Float64List matrix4}) {
     assert(path != null); // path is checked on the engine side
     assert(_offsetIsValid(offset));
-    _addPath(path, offset.dx, offset.dy);
+
```
Comment (Hixie) [inline lib/ui/painting.dart]: you can remove this blank line

File: lib/ui/painting.dart [hunk 16]
```diff
@@ -1558,21 +1586,41 @@ class Path extends NativeFieldWrapperClass2 {
...
-
+  void _addPathWithMatrix(Path path, double dx, double dy, Float64List matrix) native 'Path_addPathWithMatrix';
+  
   /// Adds the given path to this path by extending the current segment of this
   /// path with the the first segment of the given path.
-  void extendWithPath(Path path, Offset offset) {
+  /// 
+  /// If matrix4 is specified, the path will be transformed by this matrix
+  /// (after the matrix is translated by the given offset).
+  void extendWithPath(Path path, Offset offset, {Float64List matrix4}) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: same comments

File: lib/ui/painting.dart [hunk 17]
```diff
@@ -1558,21 +1586,41 @@ class Path extends NativeFieldWrapperClass2 {
...
     assert(path != null); // path is checked on the engine side
     assert(_offsetIsValid(offset));
-    _extendWithPath(path, offset.dx, offset.dy);
+
```
Comment (Hixie) [inline lib/ui/painting.dart]: remove blank line

File: lib/ui/painting.dart (hunk 14 above)
Comment (Hixie) [inline lib/ui/painting.dart]: Mention here and in the docs for any other functions that takes a Float64List matrix that the matrix argument is a 4x4 matrix, and say if it's row-major or column-major, etc.

File: lib/ui/painting.dart [hunk 18]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
   Path transform(Float64List matrix4) {
-    assert(matrix4 != null);
-    if (matrix4.length != 16)
-      throw new ArgumentError('"matrix4" must have 16 entries.');
+    assert(_matrix4IsValid(matrix4));
     return _transform(matrix4);
   }
   Path _transform(Float64List matrix4) native 'Path_transform';
+
+  /// Gets the bounding rectangle for this path.
```
Comment (Hixie) [inline lib/ui/painting.dart]: Gets -> Computes (to make it sound expensive)

File: lib/ui/painting.dart [hunk 19]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  Rect getBounds() {
+    final Float32List rect = _getBounds();
+    return new Rect.fromLTRB(rect[0], rect[1], rect[2], rect[3]);
+  }
+  Float32List _getBounds() native 'Path_getBounds';
+
+  /// Sets this path to the result of operation [PathOperation]
```
Comment (Hixie) [inline lib/ui/painting.dart]: `Combines the two paths according in the manner specified by the given operation.`

File: lib/ui/painting.dart [hunk 20]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// 
+  /// If only one path is specified, this path will be used as the first 
+  /// operand, otherwise path1 will be the first operand and path2 the
+  /// second.
```
Comment (Hixie) [inline lib/ui/painting.dart]: remove this paragraph

File: lib/ui/painting.dart [hunk 21]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// 
+  /// The resulting path will be constructed from non-overlapping contours. The
+  /// curve order is reduced where possible so that cubics may be turned into
+  /// quadratics, and quadratics maybe turned into lines.
+  static Path combine(PathOperation operation, Path path1, Path path2) {
+    assert(path1 != null);
+    assert(path2 != null);
+
```
Comment (Hixie) [inline lib/ui/painting.dart]: remove blank line

File: lib/ui/painting.dart [hunk 22]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+    final Path path = new Path();
+    if (path._op(path1, path2, operation.index)) {
+      return path;
+    } 
+    throw new StateError('Path combine failed - likely due to invalid path');
```
Comment (Hixie) [inline lib/ui/painting.dart]: Here's a more Flutter-y message:

`Path.combine() failed. This may be due to an invalid path; in particular, check for NaN values.`

File: lib/ui/painting.dart [hunk 23]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  }
+  bool _op(Path path1, Path path2, int operation) native 'Path_op';
+
+  /// Creates a [PathMetrics] object for this path.
+  PathMetrics computeMetrics({bool forceClosed = false}) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: In a second paragraph, explain what the argument does.

Also, ` = ` -> `: ` for consistency with flutter style

File: lib/ui/painting.dart [hunk 24]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+    return new PathMetrics._(this, forceClosed);
+  }
+}
+
+/// Convenience class to return the result of [PathMetric.getTangentForOffset].
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
/// The geometric description of a tangent: the angle at a point.
///
/// See also:
///
///  * [PathMetric.getTangentForOffset], which returns the tangent of an offset along a path.
```

File: lib/ui/painting.dart [hunk 25]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+class Tangent {
+  /// Position of the tangent
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
  /// The position of the tangent.
  ///
  /// When used with [PathMetric.getTangentForOffset], this represents the precise position that the given offset
  /// along the path corresponds to.
```

File: lib/ui/painting.dart [hunk 26]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  final Offset position;
+  /// Angle of the tangent in radians
```
Comment (Hixie) [inline lib/ui/painting.dart]: insert blank line between properties

Comment (Hixie) [inline lib/ui/painting.dart]:
```
  /// The direction of the curve at [position].
  ///
  /// When used with [PathMetric.getTangentForOffset], this is the angle of the curve that is the given offset
  /// along the path (i.e. the direction of the curve at [position]).
  ///
  /// This value is in radians, with 0.0 meaning pointing along the x axis in the positive x-axis direction, and
  /// positive numbers going clockwise from there.
```

Check that that part about the meaning is correct, though. :-)

File: lib/ui/painting.dart [hunk 27]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  final double angle;
+
+  const Tangent(this.position, this.angle);
```
Comment (Hixie) [inline lib/ui/painting.dart]: Please put constructors at the top of the class, before the fields. That way it's easy to tell at a glance if there's an implicit constructor without reading the whole class.

Comment (Hixie) [reply lib/ui/painting.dart]: Also the constructor should have docs, something like:

```
/// Creates a [Tangent] with the given values.
///
/// The arguments must not be null.
```

...and then add initializer asserts that verify they're not null:

```dart
  /// Creates a [Tangent] with the given values.
  ///
  /// The arguments must not be null.
  const Tangent(this.position, this.angle) : assert(position != null), assert(angle != null);
 
```

File: lib/ui/painting.dart [hunk 28]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+/// 
+/// Each path metric corresponds to a segment of a path.
+/// 
+/// For example, if you call [Path.moveTo] once in the middle of drawing, you
+/// would get two PathMetric objects in this iterable.  Call 
+/// [Path.computeMetrics] to create the iterable.
+class PathMetrics extends collection.IterableBase<PathMetric> {
+  final Iterator<PathMetric> _iterator;
+
+  PathMetrics._(Path path, bool forceClosed) :
```
Comment (Hixie) [inline lib/ui/painting.dart]: constructors first

File: lib/ui/painting.dart [hunk 29]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+}
+
+/// An iterable collection of [PathMetric] objects.
+/// 
+/// Each path metric corresponds to a segment of a path.
+/// 
+/// For example, if you call [Path.moveTo] once in the middle of drawing, you
+/// would get two PathMetric objects in this iterable.  Call 
+/// [Path.computeMetrics] to create the iterable.
+class PathMetrics extends collection.IterableBase<PathMetric> {
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
/// An iterable collection of [PathMetric] objects describing a path.
///
/// A [PathMetrics] object is created by using the [Path.computeMetrics] method,
/// and represents the path as it stood at the time of the call. Subsequent modifications
/// of the path do not affect the [PathMetrics] object.
/// 
/// Each path metric corresponds to a segment, or contour, of a path.
/// 
/// For example, a path consisting of a [Path.lineTo], a [Path.moveTo], and another [Path.lineTo]
/// will contain two contours and thus be represented by two [PathMetric] objects.
///
/// When iterating across a [PathMetrics]' contours, the [PathMetric] objects are only
/// valid until the next one is obtained.
```

File: lib/ui/painting.dart [hunk 30]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  @override
+  Iterator<PathMetric> get iterator => _iterator;
+}
+
+/// Tracks iteration from one segment of a path to the next for measurement.
+class PathMetricIterator implements Iterator<PathMetric> {
+  PathMetric _pathMetric;
+  bool _firstTime = true;
+
+  PathMetricIterator._(this._pathMetric);
```
Comment (Hixie) [inline lib/ui/painting.dart]: constructors first

File: lib/ui/painting.dart [hunk 31]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+    if (_firstTime == true) {
+      _firstTime = false;
+      return true;
+    } else if (_pathMetric?._moveNext() == true) {
+      return true;
+    } else {
+      _pathMetric = null;
+      return false;
+    }
+  }
```
Comment (Hixie) [inline lib/ui/painting.dart]: It's not a big deal but generally if your if block ends with a return, don't bother with an "else".

File: lib/ui/painting.dart [hunk 32]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+}
+
+/// Utilities for measuring a [Path] and extracting subpaths.
+///
+/// Call [Path.computeMetrics] to create these objects. Once created, measures
+/// will only be valid while the path remains unmodified.  If the path is 
+/// modified, the behavior of the PathMetric object is undefined.
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
/// Iterate over the object returned by [Path.computeMetrics] to obtain [PathMetric] objects.
///
/// Once created, metrics will only be valid while the iterator is at the given contour. When
/// the next contour's [PathMetric] object is obtained, this object becomes invalid.
￼```

File: lib/ui/painting.dart [hunk 33]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+class PathMetric extends NativeFieldWrapperClass2 {
+  /// Create a new empty [Path] object.
+  PathMetric._(Path path, bool forceClosed) { _constructor(path, forceClosed); }
+  void _constructor(Path path, bool forceClosed) native 'PathMeasure_constructor';
+
+  /// Return the total length of the current contour.
+  double get length native 'PathMeasure_getLength';
+
+  /// Pins distance to 0 <= distance <= getLength(), and then computes the 
+  /// corresponding position and tangent.
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
/// Computes the position of the current contour at the given offset, and the angle of the path at that point.
///
/// For example, calling this method with a distance of 1.4 for a line from 0.0,0.0 to 2.0,2.0 would give
/// the point 1.0,1.0 and the angle 45 degrees (but in radians).
///
/// Returns null if the contour has zero [length].
///
/// The distance is clamped to the length of the current contour.
```

File: lib/ui/painting.dart [hunk 34]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+    }
+  }
+  Float32List _getPosTan(double distance) native 'PathMeasure_getPosTan';
+
+  /// Given a start and stop distance, return the intervening segment(s).
+  /// 
+  /// startD and stopD are pinned to legal values (0..[length])
+  /// Returns null if the segment is 0 length or startD > stopD.
+  /// Begin the segment with a moveTo if startWithMoveTo is true.
+  Path extractPath(double startD, double endD, bool startWithMoveTo) native 'PathMeasure_getSegment';
```
Comment (Hixie) [inline lib/ui/painting.dart]: Rename the arguments `start` and `end`.

Make the `startWithMoveTo` argument be a named argument with a default (maybe true?).

```
/// Returns a [Path] that describes the part of this contour from `start` to `end`.
///
/// If `startWithMoveTo` is true, the default, then the returned [Path] begins with a [Path.moveTo] call.
/// If it is false, then it does not, which may be more convenient in conjunction with [Path.extendWithPath].
///
/// Returns null if the contour has zero [length] or if the `start` isn't before the `end`.
///
/// The `start` and `end` values are clamped to the [length] of the current contour.
```

File: lib/ui/painting.dart [hunk 35]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+
+  /// Returns true if the path is closed.
```
Comment (Hixie) [inline lib/ui/painting.dart]:
```
/// Whether the contour is closed.
///
/// Returns true if the contour ends with a call to [Path.close] (which may have been implied, e.g.
/// when using [Path.addRect]). Returns false otherwise.
```

Add a test that verifies that Path.addRect creates a contour that returns true here.

File: lib/ui/painting.dart [hunk 36]
```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
...
+  ///
+  /// A path can have a next contour if [Path.moveTo] was called after drawing began.
+  /// Return true if one exists, or false.
+  /// 
+  /// This is not exactly congruent with a regular [Iterator.moveNext].
+  /// Typically, [Iterator.moveNext] should be called before accessing the
+  /// [Iterator.current]. In this case, the [PathMetric] is valid before 
+  /// calling `_moveNext` - `_moveNext` should be called after the first
+  /// iteration is done instead of before.
+  bool _moveNext() native 'PathMeasure_nextContour';
```
Comment (Hixie) [inline lib/ui/painting.dart]: make these comments use `//` instead of `///` since this is private and those docs aren't "camera ready" (we'd want to rereview them before publishing them)

File: lib/ui/painting.dart (hunk 33 above)
Comment (Hixie) [reply lib/ui/painting.dart]: also, add a test that checks that if you use this method with the path described in that example, you do get those results.

**Comment by Hixie** [comment]
Thanks for your patience with this PR. I hope we'll have it landed soon! Just some style nits, some documentation changes, and if you could add some tests for this API that would be great. The tests can just run be like those here:
https://github.com/flutter/engine/blob/master/testing/dart/rect_test.dart
...but verifying the logic of Path, now that we have a way to introspect it.

**Comment by Hixie** [comment]
cc @amirh who is going to love this.

File: lib/ui/painting.dart (hunk 23 above)
Comment (dnfield) [reply lib/ui/painting.dart]: There's a note in the langauge tour suggesting that the `:` would be deprecated (assuming you're looking for `{bool forceClosed: false}`:

>Version note: Old code might use a colon (:) instead of = to set default values of named parameters. The reason is that before SDK 1.21, only : was supported for named parameters. That support is likely to be deprecated, so we recommend that you use = to specify default values, and specify an SDK version of 1.21 or higher.
https://www.dartlang.org/guides/language/language-tour#optional-parameters

Should I still change it?

File: lib/ui/painting.dart (hunk 26 above)
Comment (dnfield) [reply lib/ui/painting.dart]: Positive numbers make it point "up" towards the y axis, negative "down" - I think it's more visually helpful to reference the y-axis than to reference clock direction (counterclockwise for positive, clockwise for negative)

File: lib/ui/painting.dart (hunk 33 above)
Comment (dnfield) [reply lib/ui/painting.dart]: I'll add some tests.  The numbers won't come out quite that nice and even - what do you think of

`For example, calling this method with a distance of 1.41 ... would give a point near 1.0,1.0 and the angle near 0.785 radians (near 45 degrees)`?

**Comment by dnfield** [comment]
Thank you for all the time you've put into reviewing this - it's definitely improved for it.

I ended up running `dartfmt` on painting.dart after some changes, and it reformatted some code that I haven't otherwise touched.  Would you prefer I not use `dartfmt` on this, or should I go ahead and check in the dartfmt-ed version?

**Comment by Hixie** [comment]
You can try uploading the dartfmt version, I may have additional comments about the result. dartfmt can't handle some of the weirder things we do.

File: lib/ui/painting.dart (hunk 33 above)
Comment (Hixie) [reply lib/ui/painting.dart]: I don't think it matters if the example is 100% precise. Pedantic precision can get in the way of the message sometimes.

File: lib/ui/painting.dart (hunk 26 above)
Comment (Hixie) [reply lib/ui/painting.dart]: Hm, we should flip the sign over to match arcTo (which is documented to work as I described here).

I'm fine with using the up/down terminology if you prefer. Maybe do both. :-)

File: lib/ui/painting.dart (hunk 23 above)
Comment (Hixie) [reply lib/ui/painting.dart]: Yeah, we're using `:` everywhere for now, until we get around to switching everything over at once. Keeps things consistent in the meantime.

File: lib/ui/painting.dart [hunk 37]
```diff
@@ -1607,12 +1698,190 @@ class Path extends NativeFieldWrapperClass2 {
...
+  /// When used with [PathMetric.getTangentForOffset], this is the angle of the
+  /// curve that is the given offset along the path (i.e. the direction of the
+  /// curve at [position]).
+  /// 
+  /// This value is in radians, with 0.0 meaning pointing along the x axis in 
+  /// the positive x-axis direction, positive numbers pointing downward toward
+  /// the negative y-axis, i.e. in a clockwise direction, and negative numbers
+  /// pointing upward toward the positive y-axis, i.e. in a counter-clockwise 
+  /// direction.
+  final double angle;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting.dart]: We could probably have the best of both worlds by providing the raw vector as well as a getter that provides the angle.  e.g.

```
  final Offset tangentVector;
  double get angle => -atan2(tangentVector.y, tangentVector.x);
```

Any objections to that @Hixie ?

[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/painting.dart]: Sure, that works.

**Comment by Hixie** [comment]
This looks great! Just a couple typo fixes and then:
LGTM

**Comment by Hixie** [comment]
Thanks so much for this!

**Comment by aam** [comment]
This broke engine bots: https://github.com/flutter/engine/commit/8c00101a59cd16c30d51663b58c6f2ee9ee2e2b9#commitcomment-28495269.
